services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceHelixAdmin.java

```diff
@@ -2268,6 +2268,55 @@ private Pair<Boolean, Version> addVersion(
         null);
   }
 
+  private Optional<Version> getVersionFromSourceCluster(
+      ReadWriteStoreRepository repository,
+      String destinationCluster,
+      String storeName,
+      int versionNumber) {
+    Store store = repository.getStore(storeName);
+    if (store == null) {
+      throwStoreDoesNotExist(destinationCluster, storeName);
+    }
+    if (!store.isMigrating()) {
+      return Optional.empty();
+    }
+    ZkStoreConfigAccessor storeConfigAccessor = getStoreConfigAccessor(destinationCluster);
+    StoreConfig storeConfig = storeConfigAccessor.getStoreConfig(storeName);
+    if (!destinationCluster.equals(storeConfig.getMigrationDestCluster())) {
+      // destinationCluster is passed from the add version task, so if the task is not in the destination cluster but in
+      // the source cluster, do not replicate anything from the "source" since this is the source cluster
+      return Optional.empty();
+    }
+    String migrationSourceCluster = storeConfig.getMigrationSrcCluster();
+    ControllerClient srcControllerClient = getControllerClientMap(migrationSourceCluster).get(getRegionName());
+    if (srcControllerClient == null) {
+      throw new VeniceException(
+          "Failed to constructed controller client for cluster " + migrationSourceCluster + " and region "
+              + getRegionName());
+    }
+    StoreResponse srcStoreResponse = srcControllerClient.getStore(storeName);
+    if (srcStoreResponse.isError()) {
+      throw new VeniceException(
+          "Failed to get store " + storeName + " from cluster " + migrationSourceCluster + " and region "
+              + getRegionName() + " with error " + srcStoreResponse.getError());
+    }
+    StoreInfo srcStoreInfo = srcStoreResponse.getStore();
+    // For ongoing new pushes, destination cluster does not need to replicate the exact version configs from the source
+    // cluster; destination cluster can apply its own store configs to the new version.
```

Comment:
Is there any advantage to handling this case differently? Why not always copy from the source cluster?

Reply:
The new version might not be there in the source cluster yet, so we may or may not get it from the source; the behavior is somewhat non-deterministic.